File: solutions/maximum-depth-of-binary-tree.cpp
/*
 *Soultion of problem 104
 *https://leetcode.com/problems/maximum-depth-of-binary-tree/
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if(root == NULL)
            return 0;
        int leftMax = maxDepth(root->left);
        int rightMax = maxDepth(root->right);
        return max(leftMax, rightMax) + 1;
    }
};

File: solutions/maximum-depth-of-binary-tree.py
"""
104. Maximum Depth of Binary Tree
- https://leetcode.com/problems/maximum-depth-of-binary-tree/
"""

# Method 1
# Using DFS

class Solution(object):
    def dfs(self, node, depth):
        if not node:
            return depth - 1
        return max(self.dfs(node.left, depth + 1), self.dfs(node.right, depth + 1))

    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        return self.dfs(root, 1)
